File: Swift-Solutions/Problem-03.swift
// What is the largest prime factor of the number 600851475143 ?
import Foundation

func isPrime(num: Int) -> Bool {
  if num < 2 {
    return false
  }
  if num == 2 {
    return true
  }
  let limit: Int = Int(pow(Double(num),0.5))+1
  for i in 2...limit {
    if num % i == 0 {
      return false
    }
  }
  return true
}

func largestPrimeFactor(num: Int) -> Int {
  var maxPrimeFactor: Int = 0
  let limit: Int = Int(pow(Double(num),0.5))+1
  for i in 2...limit {
    if num % i == 0 && isPrime(num: i) {
      maxPrimeFactor = i
    }
  }
  return maxPrimeFactor
}

print(largestPrimeFactor(num: 600851475143))

File: C++ Solutions/Problem-03.cpp
/*
 What is the largest prime factor of the number 13195? 
 */

#include <iostream>
using namespace std;

bool isPrime(double num);
int largestPrimeFactor(double num);

int main(void) {
  int x = largestPrimeFactor(13195);
  cout << x << "\n";
}

bool isPrime(double num) {
  if (num <= 1) {
    return false;
  }
  if (num == 2) {
    return true;
  }
  for (int i=2;i<int(sqrt(num))+1;i++) {
    if (int(num) % i == 0) {
      return false;
    }
  }
  return true;
}

int largestPrimeFactor(double num) {
  int intNum = int(num);
  int largest = 0;
  for (int i=2;i<int(sqrt(num))+1;i++) {
    if ((intNum % i == 0) & (isPrime(i))) {
      largest = i;
    }
  }
  return largest;
}
